docs/misc/commands.md

```diff
@@ -0,0 +1,211 @@
+
+### Exceptions
+The package `com.mojang.brigadier.exceptions` provide one [exception](https://docs.oracle.com/javase/tutorial/essential/exceptions/index.html): `CommandSyntaxException`, will be thrown if the syntax is incorrect or if the command fails. It will give an error message in chat if it was a player executing the command, or as output in a [command block](https://minecraft.wiki/w/command_block) if it's a command block.
+
+It also provides many internal handler classes and the `CommandExceptionType`s.
+
+#### `CommandExceptionType`
+This is an empty interface is used to mark a class as a eponymous exception. `CommandSyntaxException` take an argument `CommandExceptionType`.
+
+The classes implementing `CommandExceptionType` aims to create exception based on a certain amount of parameters:*
+ - `DynamicCommandExceptionType`: one
+ - `Dynamic2CommandExceptionType`, `Dynamic3CommandExceptionType`, `Dynamic4CommandExceptionType`: 2, 3 or 4
+ - `DynamicNCommandExceptionType`: an undetermined amount
+ - `SimpleCommandExceptionType`: zero
+
+These classes are most used in `BuiltInExceptions` in lines like:
+```java
+private static final Dynamic2CommandExceptionType FLOAT_TOO_BIG = new Dynamic2CommandExceptionType((found, max) -> new LiteralMessage("Float must not be
+    more than " + max + ", found " + found));
+```
+and then
+```java
+public Dynamic2CommandExceptionType floatTooHigh() {
+    return FLOAT_TOO_BIG;
+}
+```
+
+:::tip
+These functions are accessible everywhere, and it's recommended to use it when throwing `CommandSyntaxException`.
+```java
+throw new CommandSyntaxException(BuiltInException.floatTooHigh(), new LiteralMessage("Please provide a smaller float!"));
+```
+:::
+
+### `SharedSuggestionProvider`
+It's the base interface of execution contexts, [client-side or server-side](../concepts/sides.md).
+
+:::warning
+We talk here about **execution** context. It's meaning that the command hasn't be written when it's instanciated, so you can't get arguments from it, use instead [`CommandContext`](#command-contexts).
+:::
+
+#### `ClientSuggestionProvider`
+It implement `SharedSuggestionProvider` but have many more function, about client context.
+
+### Command Source Stack
+The `CommandSourceStack' holds all the information about the command execution context **before** the command is written and executed.
+
+The most important ones are
+ - `boolean #hasPermission(int)`: returns whether the command executor has the given [permission level](https://minecraft.wiki/w/Permission_level).
+ - `void #sendSuccess(Supplier<Component>, boolean)`: send a success message to the executor. The supplier provides the message and the boolean if it's visible to everyone.
+ - `void #sendFailure(Component)`: send a red failure message to the executor.
+
+#### `ClientCommandSourceStack`
+It's a Neoforge class that implements `SharedSuggestionProvider`. It has no other functions, so it works like the base context.
```

Comment:
The `ClientCommandSourceStack` is meant as as a way to handle commands executed on the client rather than passed from the server.